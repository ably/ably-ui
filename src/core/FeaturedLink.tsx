import React, { CSSProperties, ReactNode } from "react";

import Icon from "./Icon";
import { ColorClass, ColorThemeSet } from "./styles/colors/types";
import cn from "./utils/cn";

type FeaturedLinkProps = {
  url: string;
  children: ReactNode;
  textSize?: string;
  iconColor?: ColorClass | ColorThemeSet;
  flush?: boolean;
  reverse?: boolean;
  additionalCSS?: string;
  newWindow?: boolean;
  onClick?: () => void;
  disabled?: boolean;
  /**
   * Optional class name for group hover state.
   */
  groupHoverClassName?: string;
};

type TargetProps = { target?: string; rel?: string };

// When generating links with target=_blank, we only add `noreferrer` to
// links that don't start with `/`, so we can continue tracking referrers
// across our own domains
const buildTargetAndRel = (url: string, newWindow: boolean) => {
  const props: TargetProps = {};

  if (newWindow) {
    props.target = "_blank";

    if (url.startsWith("/") && !url.startsWith("//")) {
      props.rel = "noopener";
    } else {
      props.rel = "noopenner noreferrer";
    }
  }

  return props;
};

const FeaturedLink = ({
  url,
  textSize = "text-p2",
  iconColor,
  flush = false,
  reverse = false,
  additionalCSS = "",
  newWindow = false,
  onClick = undefined,
  children,
  disabled = false,
  groupHoverClassName = "",
}: FeaturedLinkProps) => {
  const targetAndRel = buildTargetAndRel(url, newWindow);

  return (
    <a
      {...(onClick ? {} : { href: url })}
      className={cn(
        "font-sans font-bold block group/featured-link",
        { "text-gui-unavailable pointer-events-none": disabled },
        {
          "text-gui-default hover:text-gui-hover focus:text-gui-focus focus:outline-none focus-visible:outline-gui-focus":
            !disabled,
        },
        { "py-8": !flush },
        `ui-${textSize}`,
        additionalCSS,
      )}
      style={
        {
          "--featured-link-icon-size": `var(${textSize.replace(
            "text",
            "--fs",
          )})`,
        } as CSSProperties
      }
      {...targetAndRel}
      onClick={onClick}
    >
      {reverse ? (
        <>
          <Icon
            name="icon-gui-arrow-long-right-outline"
            size={`calc(var(--featured-link-icon-size) * 1.25)`}
            color={iconColor}
            additionalCSS={cn(
              "align-middle mr-8 relative -top-1 -right-4 transition-[right] transform rotate-180",
              { "group-hover/featured-link:right-0": !disabled },
              groupHoverClassName,
            )}
          />
          {children}
        </>
      ) : (
        <>
          {children}
          <Icon
            name="icon-gui-arrow-long-right-outline"
            size={`calc(var(--featured-link-icon-size) * 1.25)`}
            color={iconColor}
            additionalCSS={cn(
              "align-middle ml-8 relative -top-1 -left-4 transition-[left]",
              {
                "group-hover/featured-link:left-0": !disabled,
              },
              groupHoverClassName,
            )}
          />
        </>
      )}
    </a>
  );
};

export default FeaturedLink;
