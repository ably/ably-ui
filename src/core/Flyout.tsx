import React, { useState } from "react";
import {
  NavigationMenu,
  NavigationMenuItem,
  NavigationMenuList,
  NavigationMenuTrigger,
  NavigationMenuContent,
  NavigationMenuViewport,
  NavigationMenuLink,
} from "@radix-ui/react-navigation-menu";
import cn from "./utils/cn";
import { componentMaxHeight, HEADER_HEIGHT } from "./utils/heights";

/**
 * Props for the Flyout component.
 */
type FlyoutProps = {
  /**
   * Array of menu items to be displayed in the flyout.
   */
  menuItems: {
    /**
     * Name for the menu item.
     */
    name: string;
    /**
     * Optional content to be displayed in the flyout panel.
     */
    content?: React.ReactNode;
    /**
     * Optional link for the menu item.
     */
    link?: string;
    /**
     * Optional styling for the flyout panel.
     */
    panelClassName?: string;
  }[];
  /**
   * Optional class name for the flyout container.
   */
  className?: string;
  /**
   * Optional class name for the flyout element.
   */
  flyOutClassName?: string;
  /**
   * Optional class name for the menu link.
   */
  menuLinkClassName?: string;
  /**
   * Optional class name for the viewport.
   */
  viewPortClassName?: string;
};

const DEFAULT_MENU_LINK_STYLING =
  "ui-text-label3 font-bold text-ably-secondary hover:text-ably-primary px-3 py-2 flex items-center justify-between";
const DEFAULT_VIEWPORT_STYLING =
  "relative overflow-hidden w-full h-[var(--radix-navigation-menu-viewport-height)] origin-[top_center] transition-[width,_height] duration-300 data-[state=closed]:animate-scale-out data-[state=open]:animate-scale-in sm:w-[var(--radix-navigation-menu-viewport-width)]";
const PANEL_ANIMATION =
  "data-[motion=from-end]:animate-enter-from-right data-[motion=from-start]:animate-enter-from-left data-[motion=to-end]:animate-exit-to-right data-[motion=to-start]:animate-exit-to-left";

const FlyOverlay = ({
  className,
  fadingOut,
}: {
  className: string;
  fadingOut: boolean;
}) => (
  <div
    className={cn(
      "absolute left-0 right-0 h-screen w-full opacity-0",
      {
        "animate-[fade-in-ten-percent_150ms_ease-in-out_forwards]": !fadingOut,
        "animate-[fade-out-ten-percent_150ms_ease-in-out_forwards]": fadingOut,
      },
      className,
    )}
    style={{ height: componentMaxHeight(HEADER_HEIGHT), top: HEADER_HEIGHT }}
  ></div>
);

const Flyout = ({
  menuItems,
  className,
  flyOutClassName,
  menuLinkClassName,
  viewPortClassName,
}: FlyoutProps) => {
  const [isOpen, setIsOpen] = useState(false);
  const [fadingOut, setFadingOut] = useState(false);

  const closeMenu = () => {
    setFadingOut(true);

    setTimeout(() => {
      setIsOpen(false);
      setFadingOut(false);
    }, 150);
  };

  return (
    <>
      <NavigationMenu
        className={cn(className, "flex w-full")}
        onValueChange={(val) => (val ? setIsOpen(true) : closeMenu())}
        delayDuration={0}
      >
        <NavigationMenuList className="flex list-none center">
          {menuItems.map(({ name, content, link, panelClassName }) =>
            content ? (
              <NavigationMenuItem key={name}>
                <NavigationMenuTrigger
                  onClick={(event) => event.preventDefault()}
                  className={cn(
                    "group outline-none focus:outline-none select-none cursor-pointer relative",
                    "rounded-md hover:bg-ably-primary-inverse-accent",
                    "[&[data-state=open]]:bg-ably-primary-inverse-accent",
                    "[&[data-state=open]]:text-ably-primary",
                    DEFAULT_MENU_LINK_STYLING,
                    menuLinkClassName,
                  )}
                >
                  {name}
                </NavigationMenuTrigger>
                <NavigationMenuContent
                  className={cn(
                    "absolute right-0 top-0 p-6 z-10",
                    PANEL_ANIMATION,
                    panelClassName,
                  )}
                >
                  {content}
                </NavigationMenuContent>
              </NavigationMenuItem>
            ) : (
              <NavigationMenuLink
                key={name}
                href={link}
                className={cn(DEFAULT_MENU_LINK_STYLING, menuLinkClassName)}
              >
                {name}
              </NavigationMenuLink>
            ),
          )}
        </NavigationMenuList>

        <div className={cn("absolute top-full", flyOutClassName)}>
          <NavigationMenuViewport
            className={cn(DEFAULT_VIEWPORT_STYLING, viewPortClassName)}
          />
        </div>
      </NavigationMenu>
      {isOpen ? (
        <FlyOverlay
          className="bg-neutral-1300 opacity-10 z-20 h-screen mix-blend-multiply"
          fadingOut={fadingOut}
        />
      ) : null}
    </>
  );
};

export default Flyout;
