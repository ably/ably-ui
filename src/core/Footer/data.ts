import { IconName } from "../Icon/types";

type FooterLinksProps = {
  title: string;
  links: {
    label: string;
    link: string;
    badge?: string;
  }[];
};

export const footerLinks: FooterLinksProps[] = [
  {
    title: "Products",
    links: [
      { label: "Ably Pub/Sub", link: "/pubsub" },
      { label: "Ably LiveSync", link: "/livesync" },
      { label: "Ably Chat", link: "/chat" },
      { label: "Ably Spaces", link: "/spaces" },
      { label: "Ably Asset Tracking", link: "/docs/asset-tracking" },
      { label: "Compare our tech", link: "/compare" },
    ],
  },
  {
    title: "Platform",
    links: [
      { label: "Infrastructure", link: "/four-pillars-of-dependability" },
      { label: "Integrations", link: "/integrations" },
      { label: "SDKs", link: "/docs/sdks" },
      { label: "Changelog", link: "https://changelog.ably.com/" },
      { label: "Security & Compliance", link: "/security-and-compliance" },
    ],
  },
  {
    title: "Get started",
    links: [
      { label: "Documentation", link: "/docs" },
      { label: "Examples", link: "/examples" },
      { label: "Pricing", link: "/pricing" },
      { label: "Realtime A-Z", link: "/topics" },
      { label: "Support", link: "/support" },
    ],
  },
  {
    title: "Company",
    links: [
      { label: "About Ably", link: "/about" },
      { label: "Blog", link: "/blog" },
      { label: "Careers", link: "/careers", badge: "WEâ€™RE HIRING" },
      { label: "Contact us", link: "/contact" },
    ],
  },
];

export const bottomFooterLinks = [
  { label: "Data protection", link: "/data-protection" },
  { label: "Privacy", link: "/privacy" },
  { label: "Legals", link: "/legals" },
  { label: "Cookies", link: "/privacy" },
];

export const socialLinks: {
  key: string;
  colorIcon: IconName;
  monoIcon: IconName;
  link: string;
}[] = [
  {
    key: "x",
    colorIcon: "icon-social-x",
    monoIcon: "icon-social-x-mono",
    link: "https://x.com/ablyrealtime",
  },
  {
    key: "linkedin",
    colorIcon: "icon-social-linkedin",
    monoIcon: "icon-social-linkedin-mono",
    link: "https://www.linkedin.com/company/ably-realtime",
  },
  {
    key: "github",
    colorIcon: "icon-social-github",
    monoIcon: "icon-social-github-mono",
    link: "https://github.com/ably/",
  },
  {
    key: "discord",
    colorIcon: "icon-social-discord",
    monoIcon: "icon-social-discord-mono",
    link: "https://discord.gg/g8yqePUVDn",
  },
  {
    key: "youtube",
    colorIcon: "icon-social-youtube",
    monoIcon: "icon-social-youtube-mono",
    link: "https://www.youtube.com/c/AblyRealtime",
  },
];
