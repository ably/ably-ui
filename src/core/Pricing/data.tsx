import React from "react";
import { PricingDataFeature } from "./types";

export const planData: PricingDataFeature[] = [
  {
    title: {
      content: "Free",
      className: "font-mono text-p3 uppercase font-extrabold tracking-[0.16em]",
      color: "text-neutral-600 dark:text-neutral-700",
    },
    description: {
      content: "Build a proof of concept.",
      className: "ui-text-p1",
      color: "text-neutral-800 dark:text-neutral-500",
    },
    price: { amount: "$0" },
    cta: { text: "Start for free", url: "/sign-up" },
    sections: [
      {
        title: "Limits",
        items: [
          "200 concurrent channels",
          "200 concurrent connections",
          "500 messages / second",
          "6M messages / month",
        ],
      },
      {
        title: "Includes",
        items: ["Community & email support (best effort)", "No commitment"],
        listItemColors: {
          foreground: "text-neutral-700 dark:text-neutral-600",
          background: "text-neutral-300 dark:text-neutral-1000",
        },
        cta: {
          text: "See all features",
          url: "#pricing-table",
        },
      },
    ],
  },
  {
    title: {
      content: "Standard",
      className: "font-mono text-p3 uppercase font-extrabold tracking-[0.16em]",
      color: "text-neutral-1300 dark:text-neutral-000",
    },
    description: {
      content: "Roll-out into production.",
      className: "ui-text-p1",
      color: "text-neutral-800 dark:text-neutral-500",
    },
    price: {
      amount: "$29",
      content: (
        <>
          <p
            className="ui-text-p3 font-medium"
            style={{ color: "currentColor" }}
          >
            / month
          </p>
          <p className="ui-text-p2 font-bold text-gui-blue-default-dark -mt-6">
            + consumption
          </p>
        </>
      ),
    },
    cta: {
      text: "Get started",
      url: "/users/paid_sign_up?package=standard",
    },
    sections: [
      {
        title: "Limits",
        items: [
          "10k concurrent channels",
          "10k concurrent connections",
          "2.5k messages / second",
        ],
      },
      {
        title: "Includes",
        items: ["1 day email support SLA", "Uptime SLO"],
        listItemColors: {
          foreground: "text-blue-600 dark:text-blue-400",
          background: "text-blue-200 dark:text-blue-800",
        },
        cta: {
          text: "See all features",
          url: "#pricing-table",
        },
      },
    ],
  },
  {
    title: {
      content: "Pro",
      className: "font-mono text-p3 uppercase font-extrabold tracking-[0.16em]",
      color: "text-neutral-1300 dark:text-neutral-000",
    },
    description: {
      content: "Scale with confidence.",
      className: "ui-text-p1",
      color: "text-neutral-800 dark:text-neutral-500",
    },
    price: {
      amount: "$399",
      content: (
        <>
          <p
            className="ui-text-p3 font-medium"
            style={{ color: "currentColor" }}
          >
            / month
          </p>
          <p className="ui-text-p2 font-bold text-gui-blue-default-dark -mt-6">
            + consumption
          </p>
        </>
      ),
    },
    cta: { text: "Get started", url: "/users/paid_sign_up?package=pro" },
    sections: [
      {
        title: "Limits",
        items: [
          "50k concurrent channels",
          "50k concurrent connections",
          "10k messages / second",
        ],
      },
      {
        title: "Includes",
        items: ["2 hour support SLA", "Datadog (lite)", "Uptime SLO"],
        listItemColors: {
          foreground: "text-blue-600 dark:text-blue-400",
          background: "text-blue-200 dark:text-blue-800",
        },
        cta: {
          text: "See all features",
          url: "#pricing-table",
        },
      },
    ],
  },

  {
    title: {
      content: "Enterprise",
      className: "font-mono text-p3 uppercase font-extrabold tracking-[0.16em]",
      color: "text-orange-600",
    },
    description: {
      content: "Deliver without limits.",
      className: "ui-text-p1",
      color: "text-neutral-800 dark:text-neutral-500",
    },
    price: { amount: "Custom" },
    cta: {
      text: "Contact us",
      url: "/contact",
    },
    sections: [
      {
        title: "Unlimited",
        items: [
          "Concurrent channels",
          "Concurrent connections",
          "Messages / second",
        ],
      },
      {
        title: "Includes",
        items: [
          "24/7 mission critical support",
          "99.999% uptime SLAs",
          "Committed use discounts",
          "Datadog",
          "CNAME, SSO, & more",
        ],
        listItemColors: {
          foreground: "text-orange-600 dark:text-orange-600",
          background: "text-orange-200 dark:text-orange-1000",
        },
        cta: {
          text: "See all features",
          url: "#pricing-table",
        },
      },
    ],
  },
];

export const consumptionData: PricingDataFeature[] = [
  {
    title: {
      content: "Messages",
      className: "ui-text-h3",
      color: "text-neutral-1300 dark:text-neutral-000",
    },
    description: {
      content:
        "Messages contain the data that a client is communicating, such as the contents of a chat message.",
      className: "ui-text-p3",
      color: "text-neutral-700 dark:text-neutral-600",
    },
    price: { amount: "$2.50", content: "/ million" },
    sections: [
      {
        title: "Volume discounts",
        items: [
          ["Consumption", "$ / million msgs"],
          ["First 50 million msgs", "$2.50"],
          ["Next 450 million msgs", "$2.25"],
          ["Next 4.5 billion msgs", "$1.95"],
          ["Next 15 billion msgs", "$1.65"],
          ["Next 30 billion msgs", "$1.40"],
          ["Over 50 billion msgs", "$1.25"],
        ],
      },
    ],
  },
  {
    title: {
      content: "Channels",
      className: "ui-text-h3",
      color: "text-neutral-1300 dark:text-neutral-000",
      tooltip: (
        <p>
          We charge you for the amount of time a channel is active in our
          network by the minute. For example, if ten channels are in use for 45
          minutes, you will be charged a total of 450 channel minutes.
        </p>
      ),
    },
    description: {
      content:
        "Channels are used to route messages from publishers to subscribers. Channels are billed by the minute when actively being used by a connected client.",
      className: "ui-text-p3",
      color: "text-neutral-700 dark:text-neutral-600",
    },
    price: { amount: "$1.00", content: "/ million mins" },
    sections: [
      {
        title: "Volume discounts",
        items: [
          ["Consumption", "$ / million mins"],
          ["First 10 million mins", "$1.00"],
          ["Next 90 million mins", "$0.95"],
          ["Next 900 million mins", "$0.85"],
          ["Next 4 billion mins", "$0.75"],
          ["Next 10 billion mins", "$0.65"],
          ["Over 15 billion mins", "$0.60"],
        ],
      },
    ],
  },
  {
    title: {
      content: "Connections",
      className: "ui-text-h3",
      color: "text-neutral-1300 dark:text-neutral-000",
      tooltip: (
        <p>
          We charge you for the amount of time devices are connected to our
          network by the minute. For example, if ten devices are each connected
          for 45 minutes, you will be charged a total of 450 connection minutes.
        </p>
      ),
    },
    description: {
      content:
        "Clients establish and maintain a connection to the Ably service, typically over WebSockets.",
      className: "ui-text-p3",
      color: "text-neutral-700 dark:text-neutral-600",
    },
    price: { amount: "$1.00", content: "/ million mins" },
    sections: [
      {
        title: "Volume discounts",
        items: [
          ["Consumption", "$ / million mins"],
          ["First 10 million mins", "$1.00"],
          ["Next 90 million mins", "$0.95"],
          ["Next 900 million mins", "$0.85"],
          ["Next 4 billion mins", "$0.75"],
          ["Next 10 billion mins", "$0.65"],
          ["Over 15 billion mins", "$0.60"],
        ],
      },
    ],
  },
];
