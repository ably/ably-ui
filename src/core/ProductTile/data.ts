import { IconName } from "../Icon/types";

export const productNames = [
  "pubsub",
  "chat",
  "spaces",
  "liveSync",
  "assetTracking",
  "liveObjects",
] as const;

export type ProductName = (typeof productNames)[number];

type Products = Record<
  ProductName,
  {
    label: string;
    description: string;
    link?: string;
    icon?: IconName;
    hoverIcon?: IconName;
    unavailable?: boolean;
  }
>;

export const products: Products = {
  pubsub: {
    label: "Pub/Sub",
    description: "Low-level APIs for building realtime applications.",
    icon: "icon-product-pubsub-mono",
    hoverIcon: "icon-product-pubsub",
    link: "/docs/channels",
  },
  chat: {
    label: "Chat",
    description: "Build feature-rich live chat experiences.",
    icon: "icon-product-chat-mono",
    hoverIcon: "icon-product-chat",
    link: "/docs/chat",
  },
  spaces: {
    label: "Spaces",
    description: "Enhance your application with collaborative features.",
    icon: "icon-product-spaces-mono",
    hoverIcon: "icon-product-spaces",
    link: "/docs/spaces",
  },
  liveSync: {
    label: "LiveSync",
    description: "Fan-out database changes to client applications.",
    icon: "icon-product-livesync-mono",
    hoverIcon: "icon-product-livesync",
    link: "/docs/livesync",
  },
  assetTracking: {
    label: "Asset Tracking",
    description: "Track the geographic location of assets and fleets.",
    icon: "icon-product-asset-tracking-mono",
    hoverIcon: "icon-product-asset-tracking",
    link: "/docs/asset-tracking",
  },
  liveObjects: {
    label: "LiveObjects",
    description: "Sync application state across client devices.",
    icon: "icon-product-liveobjects-mono",
    hoverIcon: "icon-product-liveobjects",
    link: "/docs/liveobjects",
  },
};
