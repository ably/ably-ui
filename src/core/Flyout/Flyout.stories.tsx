import React from "react";
import Flyout from "../Flyout";

import FeaturedLink from "../FeaturedLink";
import { ProductName, products } from "../ProductTile/data";
import ProductTile from "../ProductTile";

export default {
  title: "Components/Flyout",
  component: Flyout,
  tags: ["autodocs"],
};

const platforms = [
  "Infrastructure",
  "Integrations",
  "SDKs",
  "Security & Compliance",
];

const panelClassName =
  "w-full sm:w-[816px] bg-neutral-000 dark:bg-neutral-1300";

const Panels = ({
  panelLeft,
  platforms,
}: {
  panelLeft: React.ReactNode;
  platforms: string[];
}) => (
  <div className="flex flex-row gap-x-24">
    <div className="flex-6">{panelLeft}</div>
    <div className="flex-4 pt-12">
      <p className="ui-text-overline2 text-neutral-700 dark:text-neutral-600 pb-6">
        platform
      </p>
      {platforms.map((item) => (
        <li className="list-none py-6" key={item}>
          <a
            className="ui-text-label3 text-neutral-1000 dark:text-neutral-300"
            href={`/platform/${item.toLowerCase()}`}
          >
            {item}
          </a>
        </li>
      ))}
    </div>
  </div>
);

const DefaultPanelLeft = ({ title, desc }: { title: string; desc: string }) => (
  <div className="bg-neutral-100 dark:bg-neutral-1200 w-full p-24">
    <h4 className="ui-text-h4 text-neutral-1300 dark:text-neutral-000">
      {title}
    </h4>
    <p className="ui-text-p3 text-neutral-800 dark:text-neutral-500 mt-8">
      {desc}
    </p>
    <FeaturedLink
      url=""
      additionalCSS="text-neutral-1300 dark:text-neutral-000"
      iconColor="text-orange-600"
    >
      Learn more
    </FeaturedLink>
  </div>
);

const menuItems = [
  { name: "Home", link: "" },
  {
    name: "Products",
    content: (
      <Panels
        panelLeft={
          <div className="grid grid-cols-2">
            {Object.keys(products).map((product) => (
              <ProductTile
                name={product as ProductName}
                key={product}
                animateIcons={true}
              />
            ))}
          </div>
        }
        platforms={platforms}
      />
    ),
    panelClassName: panelClassName,
  },
  {
    name: "Solutions",
    content: (
      <Panels
        panelLeft={
          <DefaultPanelLeft
            title="Fan engagement"
            desc=" Capture the attention of millions of fans during live events."
          />
        }
        platforms={platforms}
      />
    ),
    panelClassName: panelClassName,
  },
  {
    name: "Company",
    content: (
      <Panels
        panelLeft={
          <DefaultPanelLeft
            title="Leading the realtime revolution"
            desc="Hear from our founders about Ably’s ambitious plans to become the world’s definitive realtime platform."
          />
        }
        platforms={platforms}
      />
    ),
    panelClassName: panelClassName,
  },
  { name: "Pricing", link: "/pricing" },
  { name: "Docs", link: "/docs" },
];

const FlyoutStory = () => {
  return (
    <Flyout
      menuItems={menuItems}
      className="justify-center relative z-40"
      flyOutClassName="flex w-full justify-center"
      viewPortClassName="ui-shadow-lg-medium border border-neutral-000 dark:border-neutral-1300 rounded-2xl mt-8"
    />
  );
};

export const Default = {
  render: () => <FlyoutStory />,
};
