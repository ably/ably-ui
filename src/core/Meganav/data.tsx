import React from "react";
import { IconName } from "../Icon/types";
import { MeganavPanel } from "./MeganavPanel";
import Status, { StatusUrl } from "../Status";
import FanEngagementNavImage from "./images/fan-engagement-nav-image.png";
import CompanyNavImage from "./images/founders-nav-image.png";
import BestRequirementsWinter2025 from "../images/g2-best-meets-requirements-2025.svg";
import BestSupportWinter2025 from "../images/g2-best-support-2025.svg";
import HighPerformerWinter2025 from "../images/g2-high-performer-2025.svg";
import UsersMostLikelyToRecommend from "../images/g2-users-most-likely-to-recommend-2025.svg";
import { products } from "../ProductTile/data";

export type FlyoutPanelList = {
  label: string;
  icon: IconName;
  link: string;
  isMobile?: boolean;
};

export type FlyoutPanelHighlight = {
  heading: string;
  content: string;
  labelLink: string;
  url: string;
  image: string;
};

export type MenuItem = {
  name: string;
  link?: string;
  isHiddenMobile?: boolean;
  content?: React.ReactNode;
  panelClassName?: string;
};

const panelClassName = "w-full sm:w-[50.9375rem]";

const panelLeftFeatureClassName =
  "bg-neutral-100 dark:bg-neutral-1200 hidden md:grid border border-neutral-300 dark:border-neutral-1000 hover:border-neutral-400 dark:hover:border-neutral-800 rounded-lg cursor-pointer group/meganav-panel";

const productsMenu: FlyoutPanelList[] = [
  {
    label: "Infrastructure",
    icon: "icon-gui-globe-alt-outline",
    link: "/four-pillars-of-dependability",
  },
  {
    label: "Integrations",
    icon: "icon-gui-puzzle-piece-outline",
    link: "/integrations",
  },
  {
    label: "SDKs",
    icon: "icon-gui-cube-transparent-outline",
    link: "/docs/sdks",
  },
  {
    label: "Security & Compliance",
    icon: "icon-gui-shield-check-outline",
    link: "/security-and-compliance",
  },
];

const solutionsHighlight: FlyoutPanelHighlight = {
  heading: "Fan Engagement",
  content: "Capture the attention of millions of fans during live events.",
  labelLink: "Learn more",
  url: "/fan-engagement",
  image: FanEngagementNavImage,
};

const solutionsMenu: FlyoutPanelList[] = [
  {
    label: "Fan Engagement",
    icon: "icon-gui-hand-thumb-up-outline",
    link: "/fan-engagement",
    isMobile: true,
  },
  {
    label: "BizTech",
    icon: "icon-gui-building-office-outline",
    link: "/solutions/ecommerce-and-retail",
  },
  {
    label: "FinTech",
    icon: "icon-gui-currency-dollar-outline",
    link: "/solutions/fintech",
  },
  {
    label: "HealthTech",
    icon: "icon-gui-heart-outline",
    link: "/solutions/healthcare",
  },
  {
    label: "EdTech",
    icon: "icon-gui-academic-cap-outline",
    link: "/solutions/edtech",
  },
];

const companyHighlight: FlyoutPanelHighlight = {
  heading: "Leading the realtime revolution",
  content:
    "Hear from our founders about Ably’s ambitious plans to become the world’s definitive realtime platform.",
  labelLink: "About Ably",
  url: "/about",
  image: CompanyNavImage,
};

const companyMenu: FlyoutPanelList[] = [
  {
    label: "About Ably",
    icon: "icon-gui-ably-badge",
    link: "/about",
    isMobile: true,
  },
  {
    label: "Customer stories",
    icon: "icon-gui-star-outline",
    link: "/case-studies",
  },
  {
    label: "Careers",
    icon: "icon-gui-briefcase-outline",
    link: "/careers",
  },
  {
    label: "Blog",
    icon: "icon-gui-light-bulb-outline",
    link: "/blog",
  },
];

export const ablyAwards = [
  {
    image: BestRequirementsWinter2025,
    desc: "G2 Best Requirements Winter 2025",
  },
  {
    image: BestSupportWinter2025,
    desc: "G2 Best Support Winter 2025",
  },
  {
    image: HighPerformerWinter2025,
    desc: "G2 High Performer Winter 2025",
  },
  {
    image: UsersMostLikelyToRecommend,
    desc: "G2 Users Most Likely to Recommend Winter 2025",
  },
];

export const menuItemLinks = [
  { name: "Pricing", link: "/pricing", isHiddenMobile: true },
  { name: "Docs", link: "/docs", isHiddenMobile: true },
];

export const menuItemsForHeader: MenuItem[] = [
  {
    name: "Products",
    content: (
      <MeganavPanel
        displayProductTile={true}
        panelLeftClassName="grid"
        panelRightItems={productsMenu}
        panelRightHeading="platform"
        panelRightBottom={<Status statusUrl={StatusUrl} showDescription />}
      />
    ),
    panelClassName,
  },
  {
    name: "Solutions",
    content: (
      <MeganavPanel
        panelLeft={solutionsHighlight}
        panelLeftClassName={panelLeftFeatureClassName}
        panelRightItems={solutionsMenu}
      />
    ),
    panelClassName,
  },
  {
    name: "Company",
    content: (
      <MeganavPanel
        panelLeft={companyHighlight}
        panelLeftClassName={panelLeftFeatureClassName}
        panelRightItems={companyMenu}
        panelRightBottom={
          <div className="flex-1 gap-x-2 hidden md:flex">
            {ablyAwards.slice(0, 3).map((award) => (
              <img
                key={award.desc}
                src={award.image}
                alt={award.desc}
                width="57"
                height="64"
              />
            ))}
          </div>
        }
      />
    ),
    panelClassName,
  },
  ...menuItemLinks,
];

// Since the product-tile data is used in other projects, we update it here
// for the meganav.
export const productsForNav = {
  ...products,
  pubsub: { ...products.pubsub, link: "/pubsub" },
  liveSync: { ...products.liveSync, link: "/livesync" },
  chat: { ...products.chat, link: "/chat" },
  spaces: { ...products.spaces, link: "/spaces" },
  assetTracking: {
    ...products.assetTracking,
    link: "/solutions/asset-tracking",
  },
  liveObjects: { ...products.liveObjects, link: "/liveobjects" },
};
