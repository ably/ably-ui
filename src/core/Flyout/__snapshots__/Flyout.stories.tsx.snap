// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Flyout Default smoke-test 1`] = `
<nav aria-label="Main"
     data-orientation="horizontal"
     dir="ltr"
     class="justify-center relative z-40 flex w-full"
>
  <div style="position: relative;">
    <ul data-orientation="horizontal"
        class="flex list-none center"
        dir="ltr"
    >
      <a href
         class="ui-text-menu3 font-bold text-neutral-1000 dark:neutral-300 hover:text-neutral-1300 dark:hover:text-neutral-000 px-12 py-8 flex items-center justify-between"
         data-radix-collection-item
      >
        Home
      </a>
      <li>
        <button id="radix-:r0:-trigger-radix-:r1:"
                data-state="closed"
                aria-expanded="false"
                aria-controls="radix-:r0:-content-radix-:r1:"
                class="group outline-none focus:outline-none select-none cursor-pointer relative rounded-md hover:bg-neutral-100 dark:hover:bg-neutral-1200 ui-text-menu3 font-bold text-neutral-1000 dark:neutral-300 hover:text-neutral-1300 dark:hover:text-neutral-000 px-12 py-8 flex items-center justify-between"
                data-radix-collection-item
        >
          Products
        </button>
      </li>
      <li>
        <button id="radix-:r0:-trigger-radix-:r2:"
                data-state="closed"
                aria-expanded="false"
                aria-controls="radix-:r0:-content-radix-:r2:"
                class="group outline-none focus:outline-none select-none cursor-pointer relative rounded-md hover:bg-neutral-100 dark:hover:bg-neutral-1200 ui-text-menu3 font-bold text-neutral-1000 dark:neutral-300 hover:text-neutral-1300 dark:hover:text-neutral-000 px-12 py-8 flex items-center justify-between"
                data-radix-collection-item
        >
          Solutions
        </button>
      </li>
      <li>
        <button id="radix-:r0:-trigger-radix-:r3:"
                data-state="closed"
                aria-expanded="false"
                aria-controls="radix-:r0:-content-radix-:r3:"
                class="group outline-none focus:outline-none select-none cursor-pointer relative rounded-md hover:bg-neutral-100 dark:hover:bg-neutral-1200 ui-text-menu3 font-bold text-neutral-1000 dark:neutral-300 hover:text-neutral-1300 dark:hover:text-neutral-000 px-12 py-8 flex items-center justify-between"
                data-radix-collection-item
        >
          Company
        </button>
      </li>
      <a href="/pricing"
         class="ui-text-menu3 font-bold text-neutral-1000 dark:neutral-300 hover:text-neutral-1300 dark:hover:text-neutral-000 px-12 py-8 flex items-center justify-between"
         data-radix-collection-item
      >
        Pricing
      </a>
      <a href="/docs"
         class="ui-text-menu3 font-bold text-neutral-1000 dark:neutral-300 hover:text-neutral-1300 dark:hover:text-neutral-000 px-12 py-8 flex items-center justify-between"
         data-radix-collection-item
      >
        Docs
      </a>
    </ul>
  </div>
  <div class="absolute top-full flex w-full justify-center">
  </div>
</nav>
`;
